src/main/java/com/squareup/kotlinpoet/Util.kt

```diff
@@ -113,3 +113,58 @@ internal fun stringLiteralWithQuotes(value: String): String {
     return result.toString()
   }
 }
+
+internal fun isIdentifier(name: String): Boolean {
+  return IDENTIFIER_REGEX.matches(name)
+}
+
+internal fun isKeyword(name: String): Boolean {
+  return KEYWORDS.contains(name)
+}
+
+internal fun isName(name: String): Boolean {
+  return NAME_REGEX.split(name, 0).toTypedArray().none { isKeyword(name) }
+}
+
+private val NAME_REGEX = "\\.".toRegex()
+private val IDENTIFIER_REGEX
+    = ("((\\p{gc=Lu}+|\\p{gc=Ll}+|\\p{gc=Lt}+|\\p{gc=Lm}+|\\p{gc=Lo}+|\\p{gc=Nl}+)+" +
+    "\\d*" +
+    "\\p{gc=Lu}*\\p{gc=Ll}*\\p{gc=Lt}*\\p{gc=Lm}*\\p{gc=Lo}*\\p{gc=Nl}*)" +
```

Comment:
https://jetbrains.github.io/kotlin-spec/#_lexical_grammar
identifier-start (line 131) and identifier-part (132 and 133).

Yeah, it's too crazy. I'll look at it again and see if I can think of anything.